CMakeLists.txt

```diff
@@ -1,22 +1,65 @@
+    set(DFHACK ../../dfhack)
+elseif(EXISTS ../../../dfhack) # twbt is inside a directory in dfhack/plugins (probably)
+    set(DFHACK ../../../dfhack)
+else()
+    message("couldn't find a dfhack directory nearby, modify the cmake and put your dfhack path")
+    SET(DFHACK /your/path/dfhack)
+endif()
 
-STRING( REGEX REPLACE "\\." "" DV_VS "${DF_VERSION}" )
+include(${DFHACK}/CMake/Common.cmake)
+message("Version information\n-------------------")
+message("dwarf fortress: ${DF_VERSION}")
+message("dfhack:         ${DFHACK_VERSION}")
+message("twbt:           ${TWBT_VER}")
+string(REPLACE "." "" DF_VERSION ${DF_VERSION})
 
-SET( CMAKE_CXX_FLAGS "-m64 -O3 -DDFHACK_VERSION=\"${DFHACK_VERSION}\" -DDF_${DV_VS} -DTWBT_VER=\"\\\"${TWBT_VER}\\\"\"" )
+# include directories
+include_directories(${DFHACK}/library/include)
+include_directories(${DFHACK}/library/proto)
+include_directories(${DFHACK}/depends/protobuf)
+include_directories(${DFHACK}/depends/lua/include)
 
-IF( APPLE )
-  IF(CMAKE_COMPILER_IS_GNUCXX)
-    SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++0x -Wno-multichar -framework OpenGL -undefined dynamic_lookup" )
-  ELSE()
-    SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++0x -stdlib=libstdc++ -Wno-tautological-compare -framework OpenGL -mmacosx-version-min=10.6 -undefined dynamic_lookup" )
-  ENDIF()
-ELSE( APPLE )
-  SET( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++0x" )
-ENDIF( APPLE )
+# source files
+set( SRC twbt.cpp )
+set( DEP renderer.hpp config.hpp dungeonmode.hpp dwarfmode.hpp renderer_twbt.h commands.hpp plugin.hpp tileupdate_text.hpp tileupdate_map.hpp patches.hpp zoomfix.hpp buildings.hpp items.hpp units.hpp Makefile legacy/renderer_legacy.hpp legacy/twbt_legacy.hpp)
 
-DFHACK_PLUGIN( twbt ${PROJECT_SRCS} )
+find_package(PNG)
+if(NOT PNG_FOUND)
+    # XXX Probably this is wrong. It appears df-twbt/libpng has precompiled
+    # libs for Windows? Probably it should be a submodule that has the
+    # actual libpng source (which builds with CMake).
+    include_directories(libpng)
+endif()
+add_definitions(${PNG_DEFINITIONS})
+#target_link_libraries(${PLUGIN_TARGET} ${PNG_LIBRARIES})
 
-# add_subdirectory( plugins ) Already included in dfhack?
+# twbt compile options
+add_compile_options("-Wno-error") # loads of warnings
+add_compile_options("-fpermissive") # taking address of rvalue
+add_compile_definitions(DF_${DF_VERSION})
+add_compile_definitions(DFHACK_VERSION=${DFHACK_VERSION})
+add_compile_definitions(TWBT_VER=\"${TWBT_VER}\")
+
+# plugin target
+if(COMMAND dfhack_plugin)
+    dfhack_plugin(twbt ${SRC})
+else()
+    set( CMAKE_SHARED_LIBRARY_PREFIX_CXX "" )
+    set( PLUGIN_TARGET twbt.plug)
+    #add_link_options("-L${DFHACK_BUILD}/library -ldfhack -ldfhack-version")
+    link_directories("${DFHACK_BUILD}/library")
+    add_library(${PLUGIN_TARGET} SHARED ${SRC} ${DEP})
+    target_link_libraries(${PLUGIN_TARGET} PUBLIC dfhack dfhack-version)
```

Comment:
This is going to require the dfhack and dfhack-version targets to not only already be built, but to be built recently enough to be compatible with TWBT.

This is one of the bigger reasons why I think supporting only the in-DFHack-tree build approach is easier, and I imagine it's why other plugins only support it.

Reply:
if cmake knows the targets your linking it builds them if they aren't available or up-to-date already

Reply:
It can't build targets that aren't part of the current CMake project, which is the case in this branch.

Reply:
Fair, which then does require they be built before hand and present in the build directory...
Alright you know what? I yield.. I'll revise this to the bare bones situation, and update the build readme.. (correction, add one.. I'll add a build.md) I'll add the corresponding plugins sub-directory. `external` is that settled? or is `custom` preferred? I don't think you weighed in on that part

Reply:
This probably will require fixing the warnings though. And will probably make me sad when dfhack is updated for whatever the next version of DF will be, and the twbt offsets haven't been updated.. Everyone will have to move the directory or delete it else build all will fail for dfhack

Reply:
I can handle the DFHack side of it. There are other solutions to temporarily disabling a plugin, like commenting out its entry in CMakeLists.txt (and the `external` one wouldn't be tracked anyway).

I'm fine with `external`.